var/spack/repos/builtin/packages/hypre/package.py

```diff
@@ -388,31 +388,23 @@ def _cached_tests_work_dir(self):
         """The working directory for cached test sources."""
         return join_path(self.test_suite.current_test_cache_dir, self.extra_install_tests)
 
-    def test(self):
+    def test_hypre(self):
         """Perform smoke test on installed HYPRE package."""
         if "+mpi" not in self.spec:
-            print("Skipping: HYPRE must be installed with +mpi to run tests")
-            return
-
-        # Build copied and cached test examples
-        self.run_test(
-            "make",
-            ["HYPRE_DIR={0}".format(self.prefix), "bigint"],
-            purpose="test: building selected examples",
-            work_dir=self._cached_tests_work_dir,
-        )
-
-        # Run the examples built above
-        for exe in ["./ex5big", "./ex15big"]:
-            self.run_test(
-                exe,
-                [],
-                [],
-                installed=False,
-                purpose="test: ensuring {0} runs".format(exe),
-                skip_missing=True,
-                work_dir=self._cached_tests_work_dir,
-            )
+            raise SkipTest("Package must be installed with +mpi")
+
+        with test_part(self, "test_hypre_build", purpose="Build copied and cached test examples"):
+            with working_dir(self._cached_tests_work_dir):
+                exe = which("make")
+                exe("HYPRE_DIR={0}".format(self.prefix), "bigint")
+
+        for exeName in ["ex5big", "ex15big"]:
+            with test_part(self, f"test_hypre_{exeName}", purpose=f"Ensuring {exeName} runs"):
+                with working_dir(self._cached_tests_work_dir):
+                    exe = which(exeName)
+                    if not os.path.exists("./" + exeName):
```

Comment:
Good catch for skipping an executable.

To clean it up a bit, you could check for the path existing *before* calling `which()` or do what you've done in other PRs by skipping the test if `exe` is `None`.